File: cs-fundamentals/mutex-semaphore-monitor/no-mutex/no_mutex_demo.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// Shared Resource
var sharedCounter int = 0

// Function to simulate the task without Mutex
func incrementCounter(id int, wg *sync.WaitGroup) {
	// Decrease the WaitGroup counter when the function returns
	defer wg.Done()

	fmt.Printf("[Goroutine-%d] Starting task (No Lock)...\n", id)

	// --- CRITICAL SECTION (Unprotected) ---
	// Read the current value of the shared counter
	localCopy := sharedCounter

	// Simulate processing time (Context Switch usually happens here)
	// All goroutines might read the same initial value '0' before writing back
	time.Sleep(2 * time.Second)

	// Update the local copy and write back to the shared variable
	sharedCounter = localCopy + 1
	// --------------------------------------

	fmt.Printf("[Goroutine-%d] Finished. Local read: %d, Wrote: %d\n", id, localCopy, sharedCounter)
}

func main() {
	var wg sync.WaitGroup
	fmt.Println("--- Starting Simulation WITHOUT Mutex (Race Condition) ---")

	// Spawn 5 Goroutines
	for i := 1; i <= 5; i++ {
		wg.Add(1) // Increment the WaitGroup counter
		go incrementCounter(i, &wg)
	}

	// Wait for all goroutines to finish
	wg.Wait()

	fmt.Println("------------------------------")
	fmt.Printf("Expected result: 5\n")
	fmt.Printf("Actual result:   %d\n", sharedCounter)
	fmt.Println("------------------------------")
}

File: cs-fundamentals/mutex-semaphore-monitor/no-mutex/no_mutex_demo.py
import threading
import time

# Shared Resource
shared_counter = 0

# Note: No Mutex is used here

def increment_counter(thread_name):
    global shared_counter

    print(f"[{thread_name}] Starting task (No Lock)...")

    # CRITICAL SECTION (Unprotected)
    # Simulate a time-consuming task (Reading -> Processing -> Writing)
    local_copy = shared_counter
    
    # Simulate processing time (Context switch happens here)
    time.sleep(2) 
    
    # Write back the new value
    shared_counter = local_copy + 1

    print(f"[{thread_name}] Finished. Local value was {local_copy}, wrote {shared_counter}")

# Initialize threads
threads = []
print("--- Starting Simulation WITHOUT Mutex (Race Condition) ---")
for i in range(5):
    t = threading.Thread(target=increment_counter, args=(f"Thread-{i+1}",))
    threads.append(t)
    t.start()

for t in threads:
    t.join()

print("-" * 30)
print(f"Expected result: 5")
print(f"Actual result:   {shared_counter}") 
print("-" * 30)